### ISSUE TITLE
[glsl-in] Atomics support

### ISSUE THREAD

**Issue by fintelia** [gfx-rs/wgpu#4372]
Naga's GLSL frontend does not currently handle atomic operations. For example, this shader:
```glsl
#version 460 core

layout(std430, binding = 0) buffer Buffer {
    uint x;
} buf;

void main() {
    atomicAdd(buf.x, 1);
}
```
Produces this error:
```
error: Unknown function 'atomicAdd'
  ┌─ shader.comp:8:5
  │
8 │     atomicAdd(buf.x, 1);
  │     ^^^^^^^^^^^^^^^^^^^
```

(I couldn't find an existing tracking issue for this, though gfx-rs/naga#1154 and gfx-rs/naga#929 are related.)

**Comment by schell** [comment]
Would adding atomics for glsl mean that spirv gets atomics automatically?

**Comment by cwfitzgerald** [comment]
No. The frontends are generally completely separate bits of code.

**Comment by PENGUINLIONG** [comment]
Is there a plan for the GLSL atomics?

**Comment by cwfitzgerald** [comment]
The machinery in `naga` is frontend agnostic and currently only wired up within `spv-in`, it needs to to be wired up within the glsl frontend as well, there's just currently no one working on this. See `naga/src/front/atomic_upgrade.rs`

**Comment by schell** [comment]
If nobody else picks this up this year, _and_ I get another round of NLNet funding for [`renderling`](https://github.com/schell/renderling), I might be able to tackle this. Otherwise I'm always available to help guide, review, etc anybody who does want to tackle this.
